Minecraft/src/main/java/com/microsoft/Malmo/Client/VideoHook.java

```diff
@@ -316,6 +318,9 @@ public void postRender(RenderWorldLastEvent event)
             System.out.format("Failed to send frame - will retry in %d seconds\n", RETRY_GAP_NS / 1000000000L);
             retry_time_ns = time_before_ns + RETRY_GAP_NS;
             this.failedTCPSendCount++;
+        } else {
+            if (observer != null)
+                observer.frameProduced();
```

Comment:
I'm not sure this really adds a great deal, as it stands. `frameProduced` is only called if the frames send correctly, and then the code in ClientStateMachine checks that there's no timeout between sent frames. But we're already checking whether the frames sent correctly (`failedTCPSendCount`) - I'm not sure the situation will ever arise where the new check will go off before the old check? Unless Minecraft has stopped producing frames entirely, which the old check won't pick up on.

In that case, I'd suggest moving the `frameProduced()` call to near the top of this method (after the early return) - or even to the `OnRender()` method - so it literally becomes a test of whether frames are being produced, rather than produced+sent.